presets/ragengine/vector_store/base.py

```diff
@@ -34,36 +36,37 @@ def generate_doc_id(text: str) -> str:
         """Generates a unique document ID based on the hash of the document text."""
         return hashlib.sha256(text.encode('utf-8')).hexdigest()
 
-    def index_documents(self, index_name: str, documents: List[Document]) -> List[str]:
+    async def index_documents(self, index_name: str, documents: List[Document]) -> List[str]:
         """Common indexing logic for all vector stores."""
         if index_name in self.index_map:
-            return self._append_documents_to_index(index_name, documents)
+            return await self._append_documents_to_index(index_name, documents)
         else:
-            return self._create_new_index(index_name, documents)
+            return await self._create_new_index(index_name, documents)
 
-    def _append_documents_to_index(self, index_name: str, documents: List[Document]) -> List[str]:
+    async def _append_documents_to_index(self, index_name: str, documents: List[Document]) -> List[str]:
         """Common logic for appending documents to existing index."""
         logger.info(f"Index {index_name} already exists. Appending documents to existing index.")
         indexed_doc_ids = set()
 
         for doc in documents:
             doc_id = self.generate_doc_id(doc.text)
-            if not self.document_exists(index_name, doc, doc_id):
-                self.add_document_to_index(index_name, doc, doc_id)
+            retrieved_doc = await self.index_map[index_name].docstore.aget_ref_doc_info(doc_id)
```

Comment:
I suggest you to use asyncio.gather multi coroutines for better performance when you kick off coroutine in a loop. 
Take this page for reference. https://realpython.com/async-io-python/#chaining-coroutines

Reply:
awesome added